[aot_inductor] added runtime checks for input/output tensors in debug compile mode

chenyang78 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #122047

This PR added runtime checks to guard the dtypes and shapes of input/output tensors.
Currently, we enable these only for debug compilation
(i.e. aot_inductor.debug_compile is True) in abi_compatible mode.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

Differential Revision: [D54993148](https://our.internmc.facebook.com/intern/diff/D54993148)

Commit:
[aot_inductor] added runtime checks for input/output tensors in debug compile mode

This PR added runtime checks to guard the dtypes and shapes of input/output tensors.
Currently, we enable these only for debug compilation
(i.e. aot_inductor.debug_compile is True) in abi_compatible mode.

[ghstack-poisoned]

chenyang78 commented:
@chenyang78 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D54993148).

desertfire commented on torch/_inductor/codegen/cpp_wrapper_cpu.py:
```diff
@@ -253,6 +257,92 @@ def get_input_cpp_type(input):
+        def gen_check(handle_kind, idx, name, tensor):
+            self.prefix.writeline(f"auto {name} = {handle_kind}[{idx}];")
+            self.codegen_tensor_dtype_name_var_decl(self.prefix, name)
+            expected_dtype = dtype_to_name[tensor.dtype]
```
I think we can check dtype by comparing enum values, not by strcmp.

chenyang78 replied:
I tried to find an aten API to get the enum value for each dtype in C++ world, but couldn't. Particularly, we have aten tensor's ```dtype``` returns a ```TypeMeta``` object, which doesn't seem to provide an API to obtain the enum value for it:

```
https://caffe2.ai/doxygen-c/html/classat_1_1_tensor.html

https://caffe2.ai/doxygen-c/html/classcaffe2_1_1_type_meta.html
```

desertfire replied:
I think `dtype().toScalarType()` should do it.

chenyang78 replied:
Ah, Thanks! It turned out we actually have an existing shim function ```aoti_torch_get_dtype```. I ended up using this one.

chenyang78 commented on torch/_inductor/codegen/cpp_wrapper_cpu.py:
```diff
@@ -253,6 +257,92 @@ def get_input_cpp_type(input):
         return f"ArrayRefTensor<{DTYPE_TO_CPP[input.get_dtype()]}>"
 
+    def generate_input_output_runtime_checks(self):
+        dtype_to_name = {
```
That was my first try, but it turned out we cannot. The dtype names produced from aten API are different from those that we use for cpp declarations.

chenyang78 commented on torch/_inductor/codegen/cpp_wrapper_cpu.py:
```diff
@@ -253,6 +257,92 @@ def get_input_cpp_type(input):
+            )
+            self.codegen_input_size_var_decl(self.prefix, name)
+            for dim_idx, d in enumerate(tensor.get_size()):
+                if not isinstance(d, (int, sympy.Integer)):
```
Good point! Will fix it. Thanks.

Commit:
Update on "[aot_inductor] added runtime checks for input/output tensors in debug compile mode"


This PR added runtime checks to guard the dtypes and shapes of input/output tensors.
Currently, we enable these only for debug compilation
(i.e. aot_inductor.debug_compile is True) in abi_compatible mode.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

Differential Revision: [D54993148](https://our.internmc.facebook.com/intern/diff/D54993148)

[ghstack-poisoned]

chenyang78 commented:
@chenyang78 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D54993148).

Commit:
Update on "[aot_inductor] added runtime checks for input/output tensors in debug compile mode"


This PR added runtime checks to guard the dtypes and shapes of input/output tensors.
Currently, we enable these only for debug compilation
(i.e. aot_inductor.debug_compile is True) in abi_compatible mode.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

Differential Revision: [D54993148](https://our.internmc.facebook.com/intern/diff/D54993148)

[ghstack-poisoned]

chenyang78 commented:
@chenyang78 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D54993148).

desertfire commented on torch/_inductor/codegen/cpp_wrapper_cpu.py:
```diff
@@ -253,6 +254,123 @@ def get_input_cpp_type(input):
+        def gen_check(handle_kind, idx, name, tensor):
+            self.prefix.writeline(f"auto {name} = {handle_kind}[{idx}];")
+            self.codegen_tensor_dtype_var_decl(self.prefix, name)
+            expected_dtype, expected_dtype_name = dtype_to_name[tensor.dtype]
```
You can use `aoti_torch_dtype_*` to get `expected_dtype`, instead of creating another dict store the mapping. For  `expect_dtype_name`, I think it makes more sense to read from https://github.com/pytorch/pytorch/blob/88ebdbc97c103271766203df6662240e95a09b42/torch/_inductor/codegen/cpp.py#L83

chenyang78 replied:
> You can use aoti_torch_dtype_* to get expected_dtype, instead of creating another dict store the mapping.

Hmm, if my understanding is correct, I think you are suggesting that we construct something like ```aoti_torch_dtype_{tensor.dtype}``` based on the dtype of the tensor? Seems it doesn't work, because we want to query the runtime dtype form the real input tensor instead of the tensor from the compilation time. In other words, we don't know the dtype of the real input at the runtime, and this is actually the purpose of inserting runtime checks - we want to check this "unknown" dtype is the same as the one used at the compilation time.

> For expect_dtype_name, I think it makes more sense to read from DTYPE_TO_ATEN

Good suggestion. Fixed. Thanks!

desertfire replied:
Isn't `{name}_dtype` what you get as the real input dtype? I think expected_dtype is statically known? The reason I am suggesting that is because mapping name to a enum value can be fragile (although the dtype enum values are unlikely to change in reality).

chenyang78 replied:
Oh, I see. You were suggesting we use ```aoti_torch_dtype_``` to retrieve the expected dtype. That's a good idea. Fixed. Thanks.

desertfire reviewed (commented):
The current implementation is fine as the first step. A more holistic solution is to work with _AddRuntimeAssertionsForInlineConstraintsPass in Export, cc @angelayi

angelayi commented:
> The current implementation is fine as the first step. A more holistic solution is to work with _AddRuntimeAssertionsForInlineConstraintsPass in Export

If I'm understanding the PR correctly, this seems to be adding assertions to check just the input/output tensors which is similar to export's pass [here](https://www.internalfb.com/code/fbsource/fbcode/caffe2/torch/_export/utils.py?lines=24). However, the pass doesn't modify the graph -- it just checks the inputs during eager execution. _AddRuntimeAssertionsForInlineConstraintsPass adds assertions into the graph for inline constraints due to some torch._check or torch._check_is_size call.

desertfire commented:
> > The current implementation is fine as the first step. A more holistic solution is to work with _AddRuntimeAssertionsForInlineConstraintsPass in Export
> 
> If I'm understanding the PR correctly, this seems to be adding assertions to check just the input/output tensors which is similar to export's pass [here](https://www.internalfb.com/code/fbsource/fbcode/caffe2/torch/_export/utils.py?lines=24). However, the pass doesn't modify the graph -- it just checks the inputs during eager execution. _AddRuntimeAssertionsForInlineConstraintsPass adds assertions into the graph for inline constraints due to some torch._check or torch._check_is_size call.

So we need a pass to insert checks into the graph so that Inductor can see them. Again, it doesn't seem high-pri to me.

Commit:
Update on "[aot_inductor] added runtime checks for input/output tensors in debug compile mode"


This PR added runtime checks to guard the dtypes and shapes of input/output tensors.
Currently, we enable these only for debug compilation
(i.e. aot_inductor.debug_compile is True) in abi_compatible mode.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

Differential Revision: [D54993148](https://our.internmc.facebook.com/intern/diff/D54993148)

[ghstack-poisoned]

desertfire commented on torch/_inductor/codegen/cpp_wrapper_cpu.py:
```diff
@@ -253,6 +254,126 @@ def get_input_cpp_type(input):
+
+            self.codegen_input_stride_var_decl(self.prefix, name)
+            for stride_idx, s in enumerate(tensor.get_stride()):
+                if not isinstance(s, (int, sympy.Integer)):
```
I think you hit some strides miss-match before. I wonder if that case would be caught if we skip symbolic stride here.

chenyang78 replied:
Yeah, that particular case would be caught because the unmatched stride wasn't symbolic.

Commit:
Update on "[aot_inductor] added runtime checks for input/output tensors in debug compile mode"


This PR added runtime checks to guard the dtypes and shapes of input/output tensors.
Currently, we enable these only for debug compilation
(i.e. aot_inductor.debug_compile is True) in abi_compatible mode.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

Differential Revision: [D54993148](https://our.internmc.facebook.com/intern/diff/D54993148)

[ghstack-poisoned]

Commit:
Update on "[aot_inductor] added runtime checks for input/output tensors in debug compile mode"


This PR added runtime checks to guard the dtypes and shapes of input/output tensors.
Currently, we enable these only for debug compilation
(i.e. aot_inductor.debug_compile is True) in abi_compatible mode.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

Differential Revision: [D54993148](https://our.internmc.facebook.com/intern/diff/D54993148)

[ghstack-poisoned]

chenyang78 commented:
@pytorchbot merge

chenyang78 commented:
@pytorchbot merge

Commit:
Update on "[aot_inductor] added runtime checks for input/output tensors in debug compile mode"


This PR added runtime checks to guard the dtypes and shapes of input/output tensors.
Currently, we enable these only for debug compilation
(i.e. aot_inductor.debug_compile is True) in abi_compatible mode.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

Differential Revision: [D54993148](https://our.internmc.facebook.com/intern/diff/D54993148)

[ghstack-poisoned]

chenyang78 commented:
@pytorchbot merge
